[WIP] Code quality, browserify, es6, tooling

n8rzz opened the pull request:
Opening this PR for visibility with #678 .  This branch is still very much under development. 

Changes contained herein are massive and will cause breakages for anybody currently in development.  This would be for a v3.0.0 release due to the massive changes in the codebase.

Major items addressed or being addressed in this MR:
- [x] adds `.editorconfig`
- [x] adds `npm` support and `node_modules` for external dependency management
- [x] adds project information to `package.json` file
- [x] adds build tasks: `gulp build` and `gulp dist` (and many more supporting tasks) to transpile, concat, minify, etc source files into the `assets` dir.
- [x] adds `src` dir for development
- [x] adds support for ES6/ES2015 via babel
- [x] adds `eslint` and `eslint-plugin-airbnb` ruleset
- [x] moves script import to foot of `index.html`
- [x] updates script to only import `bundle.js`
- [x] removes need for `load_modules` function in `modules.js` and removes scripted insertion of `script` tags
- [x] manually instantiates each file/class previously located inside the `MODULES` array.
- [x] adds callbacks in appropriate places for each module file
- [x] manually adds some functions to `window` for use elsewhere
- [x] begins addressing lint errors/warnings (eslint found 14,494 total errors/warnings)
- [x] add documentation for build tools
- [x] add `ava` and `nyc` for testing and reporting 
- [x] abstract common modules across application into self-contained units
- [x] remove need for `Fiber`dependency in favor of plain es6 inheritance.
- [x] add `npm run test` task
- [x] translate files to ES6 where applicable
- [x] separate out concerns across smaller files. (This is a biggie)
- [x] `util.js` needs to be split into smaller files grouped by concern

---

Upcoming tasks (in no particular order):
- Refactor/simplify `AircraftInstanceModel`, `AircraftFlightManagementSystem`, `Leg` and `Waypoint` classes.
- remove global vars from the window, explicitly pass them where necessary (WIP)
- Explicitly pass window classes as parameters to only the class(es) that need them.
- integrate `lodash` where applicable (WIP) 
- update `for/in` loops to use either `for` or `_map` when iterating through arrays.  for reasons why, see: http://stackoverflow.com/questions/5263847/javascript-loops-for-in-vs-for (WIP) 
- simplify `if/else if/else` logic blocks.  Some methods/functions are 500+ lines long. There is room for abstraction and simplification there. (WIP) 
- Various other abstractions and simplifications have been called out via `TODO` and `FIXME` comments and will be addressed soon.
- add tests

I'll add to/update this list as I continue to make progress.  Feel free to add comments and feedback as you see fit.  I didn't open this MR just to look at it! 😄

n8rzz committed:
Adds .editorconfig file

n8rzz committed:
Adds npm and package.json
* Adds src and build dirs
* Adds paths.js to tools for consumtion by gulp tasks
* Adds initial gulp packages and tasks

n8rzz committed:
Adds concat, minify and autoprefixer to styleTasks
* Adds gulp copy tasks for images and fonts
* Adds clean tasks for images and fonts

n8rzz committed:
Removes unused gulp task.

n8rzz committed:
Adds copy and clean task for json data files

n8rzz committed:
Merge branch 'feature/npm-gulp' into develop

n8rzz committed:
Implements temp copy task to move scripts from src -> build
* Updates prev/next img refs to build dir
* Updates index.html src for javascript imports

n8rzz committed:
Adds several new packages for gulp tasks and updates scriptTasks.

n8rzz committed:
Begins moving files out of .load_module() .call_module() functions in
preperation for a full changeover to browserify

n8rzz committed:
Removes assets dir
* Adds watch:scripts task
* Adds .jshintrc file in prep for jshint
* Moves airline.js out of load_modules

n8rzz committed:
Moves tutorial out of load_modules
* Starts addressing syntax issues, mostly if/else statements and indentation

n8rzz committed:
Moves speec module out of load_modules
* Updates loadAssets function in several files to pull from build/assets
* Adds copy task for geojson files

n8rzz committed:
Moves input out of load_modules. Removes jquery and fiber files from scripts, they are now being imported and are not needed.

n8rzz committed:
Moves game out of load_modules and rearranges method calls in modules.js

n8rzz committed:
Moves ui out of load_modules.

n8rzz committed:
Removes load from load_modules

n8rzz committed:
Moves canvas out of load_modules

n8rzz committed:
Moves base out of load_modules

n8rzz committed:
Moves get out of load_modules

n8rzz committed:
Moves uitl out of load_modules

n8rzz committed:
Moves animation out of load_modules

n8rzz committed:
Finishes moving module files to a vanilla require import, thereby
allowing for browserify to bundle all scripts into one single file.

n8rzz committed:
Update version number in package.json

n8rzz committed:
Adds babel, eslint, eslint-airbnb and supporting pkgs
* Adds tasks for babel and lint
* Consolidates build and dist tasks
* Removes build dir and reverts back to assets dir as the build destination

n8rzz committed:
Pulls in tools changes from feature/clean-the-window branch

n8rzz committed:
Removes build/ dir
* abstracts time() to its own file
* addresses lint errors in module.js
* adds stub App.js file as a future entry point to the application

n8rzz committed:
Adds imports to animation file
* Corrects lint errors in animation.js

n8rzz committed:
Adds ava and nyc for tests and reports
* Adds unitConverters file with simple unit conversions
* Adds tests for new unitConverters file
* Imports unitConverters functions where needed

n8rzz commented:
@erikquinn  So the above list is accurate to what I have currently merged in to my `develop` branch.  However, I am actively developing on this branch, If you'd like to take a peak: https://github.com/n8rzz/atc/tree/feature/clean-the-window.

The meat of this work is the 1st and second items of my todo list above.  I decided to start with the `aircraft.js` file because, wow, _sooo_ many lines.

n8rzz committed:
Breaks up aircraft.js into smaller files
 * Updates imports for newly split up files and removes unused refs in aircraft.js

n8rzz committed:
Addresses most lint errors in aircraft.js
* Adds math/ with distance.js and vector.js for future math calc abstrations
* Adds happy-path test files for math/
* Updates refs for abstracted math functions.

n8rzz committed:
Furthur abstracts some util functions to specific math/
* Streamlines function logic in aircraft.js
* Adds new helper methods to flightMath/
* Adds new tests
* Updates various imports for new helpers

n8rzz committed:
At a stopping point with aircraft.js, will need to finish work on other
files before I can come back.
* Adds documentation to flightMath
* Begins work on AircraftConflict

n8rzz committed:
Merge branch 'feature/clean-the-window' of https://github.com/n8rzz/atc into develop

erikquinn commented:
@n8rzz I was hoping we might be able to push the upgrades live somewhat incrementally.... if it does have to be kind of "everything at once", perhaps it's best to put this work on a separate branch eg `zlsa/v3` which would eventually be merged into `zlsa/gh-pages`?  

That way, we are all able to test locally and find and make whatever small fixes and amendments that might be needed before pushing it live to the main branch of the game. This way, it doesn't all have to be 100% perfect in a single supermassive PR, because that could take quite a while to do (I would think). Thoughts?

n8rzz committed:
Create airport/ dir
* Creates Arrival/ dir and moves individual arrival types to thier own file
* Creates Departure/ dir and moves individual departure types to thier own file
* Move Runway and Airport to thier own files

n8rzz commented:
@erikquinn Yeah, sorry about that.  I kinda just went for it.  But, now that I think of it, it would probably be best to do all at once.  There are a lot of changes and I'd hate to affect any other development.  An all at once plan will be a little pain once instead of a little pain many times.

I think a `zlsa/v3` branch would be the best plan anyway.  With a refactor this size (I have or will have touched essentially every line of js), I am expecting at least _some_ bugs.  I've been trying to carefully test everything as I go, but I don't know the app as well as the rest of you.  So I'm sure I will have missed something.  I would feel much better knowing it's been kicked around a little bit before being released out into the wild.

Good news is with the new file structure, finding and fixing bugs should be significantly easier.

---

The thing that always gets me about refactors is this: If I've done a good job, the end user will hardly notice a difference.  But the code that runs it all looks completely different.  Crazy.

n8rzz committed:
Moves Airport factories to own files
* Mve imports to specific factory file
* Rearragnes functions in airport.js to remove the need for hoisting

erikquinn commented:
> But, now that I think of it, it would probably be best to do all at once. There are a lot of changes and I'd hate to affect any other development. An all at once plan will be a little pain once instead of a little pain many times.

^ My thoughts exactly: if we're going to do it, let's just put the rest on pause, do it RIGHT once and for all.

> I am expecting at least some bugs. I've been trying to carefully test everything as I go, but I don't know the app as well as the rest of you.

^ Oh, for sure, but the benefits _immeasurably_ outweigh the potential bugs, as it would make it so much easier to work with for those of us working on implementing improvements to the "meat and bones" of the simulator. I'll make some time to look through what you've done and make comments, ask questions, etc so we can all be on the same page. And once it's workable, we can post a link for people to test it on the web instead of having to check out a local copy.

> The thing that always gets me about refactors is this: If I've done a good job, the end user will hardly notice a difference. But the code that runs it all looks completely different. Crazy.

^ That's the idea! :wink: 

Thanks a lot for taking the time to offer up all these improvements; they will make a really big difference. I think I speak for everybody when I say your help is quite welcome! :+1:

erikquinn commented:
Added branch [`zlsa/v3`](https://github.com/zlsa/atc/tree/v3), and made it the target branch for this PR.

n8rzz committed:
Adds lodash _clamp to app
* Renames AirportInstance -> AirportInstanceModel
* Begins fixing lint errors in AirportInstanceModel.js

n8rzz committed:
Removed build/ dir from version control.

n8rzz committed:
Removes gulp tasks for copying json data.

n8rzz committed:
Fixes spacing and syntax issues with AirportInstanceModel
* replaces .hasOwnProperty with _has()
* Replaces some other operations with lodash functions
* Simplifies some small logic blocks

n8rzz committed:
Updates Runway with proper spacing and adds notes.

n8rzz committed:
Addds style/ and script/ to version control

n8rzz commented:
@erikquinn I'm making good progress here.  

Here is the loading profile for the current site:
<img width="293" alt="screen shot 2016-09-11 at 2 55 42 pm" src="https://cloud.githubusercontent.com/assets/3325985/18420071/fa83324c-782f-11e6-8a26-bbb5d93e583c.png">

and here is the loading profile for my `develop` branch:
<img width="301" alt="screen shot 2016-09-11 at 2 55 58 pm" src="https://cloud.githubusercontent.com/assets/3325985/18420083/0bfb2afc-7830-11e6-9b85-6a790bbf997d.png">

With the changes I've made so far, load time has decreased by over a second.  Ship it!

---

I look forward to your feedback!  I have tried to leave a lot of breadcrumbs along the way as I travel through the app.  There are a lot of spots that can be improved, some that I'm doing now and some that will take a little more time.  I've noticed a very high frequency of `hasOwnProperty` checks, is there any reason for these?  I'm not sure these are all entirely necessary and may be better accomplished with initializations so at least you know the property exists.

The process of a refactor often starts with a sledge hammer approach; smash big bits of code and let the pieces fall where they may.  I'm taking more of a surgical approach, working on smaller sections at a time and making sure my changes won't have side effects.  I am trying to lay the groundwork, though, for future smaller refactors like moving the functions in the `util` folder to more appropriate folders and also adding tests for those functions.  I've started doing that in some areas, but those are smaller changes that I think can wait until the big stuff is done.

To be honest, I've spent most of my time fixing the spacings and adding curly brackets to `if`s and `for`s.  Curly brackets always, it just makes things easier to ready and reason about.  It also doesn't rely on spacing to be interpreted correctly.  This isn't yaml after all, lol!

Ok, back to work...

erikquinn commented on .editorconfig:
```diff
@@ -0,0 +1,18 @@
+
+[*]
+indent_style = space
+indent_size = 4
```
I don't know others' takes on this, but as for me _personally_, I prefer 2-space indentation. What would you think about everything to tabbed, and then individuals can specify in their IDEs the size of the tabs? Consistent in the codebase, and still accommodates personal preferences.

n8rzz replied:
Ahh the old tabs vs spaces, 2 vs 4 debate.  

As you can tell, I subscribe to the 4 spaces approach.  For three reasons:
- Its just old habit.  Its what I use at work everyday and it's what I'm used to.
- I would argue that 4 makes blocks easier to pick out at a glance and increases readability.
- And I choose spaces because it just looks better in atom/vim.

With that being said, it is a preference and this isn't my codebase so its not really up to me.  I will say that all the files I've converted, I have converted to 4 spaces.  It also appears that the original version were also in spaces.  So I just changed it to 4 for readability.  

If your curious, our company JS standards are in a repo. I more or less keep to these standards: https://github.com/thenerdery/javascript-standards.
We also have HTML/CSS standards here: https://github.com/thenerdery/html-css-standards

erikquinn replied:
I see. Well it's not mine either, I just make contributions here and there. But the 4 spaces seems to be more common anyway, so I don't mind, especially since there can easily be others who cringe at 2-space indentation :wink: Thanks for the nerdery links! Looks like some good reading.

erikquinn commented on Gulpfile.js:
```diff
@@ -0,0 +1,40 @@
+/* eslint-disable */
```
ref. all the various task runners-- does gulp have a native web server? I've only used it very little. For that, I know webpack and webpack-dev-server is quite nice. At the end of the day, it doesn't matter much either way, but thought I might bring it up.

n8rzz replied:
Gulp does not have one built in, but it would be trivial to add a command to fire one up.  Personally, I use [http-server](https://www.npmjs.com/package/http-server) when developing locally.  I could very easily wire up an `npm run start` task that would fire up a server.

erikquinn commented on assets/images/icons/airport.svg:
```diff
@@ -1,95 +1 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Wouldn't we want to keep these svg's? In case we change anything with the theme/colors, etc, or want to make small changes to stuff. I'd be inclined to just leave them in there, even if we don't really need them for much of anything.

n8rzz replied:
no worry, the svgs are all still there.  The diff is going to be hard to read for this MR, due to all the changes.  I did add an `imagemin` task that will optimize the pngs and svgs, but I'm not sure that is actually necessary.

erikquinn commented on src/assets/scripts/aircraft/AircraftConflict.js:
```diff
@@ -0,0 +1,268 @@
+import Fiber from 'fiber';
+import { vlen, vsub } from '../math/vector';
```
:+1: for doing it like this. Much more organized.

n8rzz replied:
yeah, it's much cleaner.  It will take some time to move everything over to this style.  So some of that stuff will have to wait until the heavy lifting is out of the way.

n8rzz committed:
Reverted changes to parseTerrain method in AirportInstanceModel. Refactor here cause terrain data to not be drawn on the canvas.

n8rzz committed:
Updates ArrivalBase spacing and syntax
* Adds logLevel constants file and implements LOG across app

n8rzz committed:
Fixes spacing and syntax with the rest of the Arrival files

n8rzz committed:
Updates spacing in Departure files
* Removes unused imports in Arrival and Departure files
* Updates temp eslint-disabled rules in Arrival and Departure files
* Turns off fonts and images tasks for gulp dist
* Adds tau() for Math.PI * 2 and adds test

n8rzz commented:
I have removed the fonts and images task from `gulp dist`.  They are still available as a  tool, but will no longer run with that specific command.

n8rzz committed:
Updates spacing, syntax and addresses several lint errors in AircraftFlightManagementSystem.js
* Begins adding constants to AircraftConflict.js to be abstracted alter
* Updates imports in AirportInstanceModel and AircraftInstanceModel

n8rzz committed:
Stopping point for AircraftInstanceModel updates
* Adds constants for flight modes, flight category, waypoint nav mode, and classnames. These are temporarily included in this file and will likely be abstracted ata later date.
* Fixes lint errors
* Simplifies some logic, adds template strings

n8rzz committed:
Finished spacing and syntax updates with AircraftInstanceModel

n8rzz committed:
Updates AircraftModel, Leg and Waypoint syntax errors.

n8rzz committed:
Moves airline.js to airline/airline.js
* Fixes syntax and linting issues

n8rzz committed:
Moves game to game/game.js
* Creates selectors/ and begins adding selectors constants
* Stubs in storageKeys constants

n8rzz committed:
Fixes bug with .choose_weight() function
* Removes dead code in modules.js
* Removes bundle from vc
* Stubs in module methods in App.js

n8rzz committed:
Updates ui.js syntax errors
* Adds selectors file for dom manipulations
* Replaces string selector values with constants
* Adds storageKeys constants file

n8rzz committed:
Removes $ from the window and explicilty imports where needed
* Moves tutorial to own dir
* Add TutorialStep class
* Adds constant selectors to ui.js

n8rzz committed:
Updates tutorial syntax errors and formating
* Adds lint table output for gulp lint:scripts tast
* Updates eslintrc to use babel-parser

n8rzz committed:
Minor updates accross files.  Begins fixing syntax and lint errors in util.js before those functions get moved to more appropriate files.

n8rzz committed:
Adds README for tooling in the `tools/` folder
* Removes lint task and adds coverage task
* Adds author and contributor information to package.json

n8rzz committed:
Adjusts function order in ui.js to avoid the need for hoisting.

n8rzz committed:
Starts making syntax updates to util.js.  Updates author blocks in package.json

n8rzz committed:
Finishes major syntax fixes in util.js

n8rzz committed:
Fixes syntax and lint errors with input.js
* replaces strings with constants where able
* Adds new items to selectors file

n8rzz committed:
Updates references and imports for core math functions.

n8rzz committed:
Adds storageKey constants to airport.js, adds more classnames and ids to constants/selectors file.

n8rzz committed:
adds .idea/ to .gitignore

n8rzz committed:
Fixes departure route bug where route was being drawn as a straight line from runway.

n8rzz committed:
Updates cavas.js syntax and spacing.  Pulls FLIGHT_MODES out to own constants file.

n8rzz committed:
Adds constants and string interpolation to canvas.js

n8rzz committed:
Replaces $.each with _forEach and addresses double equal issue on number strings, canvas:1093

n8rzz committed:
Updates syntax and spacing in load.js

n8rzz committed:
Renames load.js to LoadingView.js
* Completly re-writes load.js to a class enstead of an IIEF
* Updates references in module.js
* Removes LoadUI from window

n8rzz committed:
Adds new app entry point at index.js -> App.js
* Begins refactoring modules.js to App.js
* Begins centralizing App configurations `window.prop` and `window.atc`

n8rzz committed:
Finishes module.js abstraction in prep for removing window methods.

n8rzz committed:
Pulls speech.js functions off the window
* Replaces strings with defined constants
* Updates refs accross files that need access to speech functions

n8rzz committed:
Fixes typo in speech.js

n8rzz committed:
Fixes bug with planes being passed to Center immediately after to clearence was issued.

n8rzz committed:
Merge branch 'develop' into feature/move-to-app

n8rzz committed:
Fixes bug with restricted areas not rendering

n8rzz committed:
Fixes bug in the Airport list where the active class remains on unselected airports.

n8rzz committed:
Modifies ui functions to arrow functions and adds export to functions used by other scripts.

n8rzz committed:
Updates ui_log to use strict paramaters instead of pulling out of `aurguments` and simplifies function logic

n8rzz committed:
Reorders functions to more closely follow call order and updates some minor syntax issues.

n8rzz committed:
Abstract GameOptions to an es6 class
* Removes Fiber dep for GameOptions
* Abstract individual game options to new file, imported by GameOptions class

n8rzz committed:
Updates looping logic in ui
* Moves prop to this.prop as a first step away from window.prop

n8rzz committed:
Converts Runway to a standard es6 class

n8rzz committed:
Updates AirportInstance and all Arrival files to use es6 classes. Removes Fiber dependency.

n8rzz committed:
Converts Departure files to es6 classes.

n8rzz committed:
Converts Airline to es6 class.

n8rzz committed:
Adds super calls in Arrival files
* Continues move away from fiber

n8rzz committed:
Adds AirlineController and abstracts window methods to be encapsulated in this new controller

n8rzz committed:
Moves aircraft.js to AircraftController
* Moves all aircraf_* functions to AircraftController and exposes only a single instance of AircraftController to the window
* Updates aircraft function references throughout app

n8rzz committed:
Moves airport.js to AirportController
* Consolidates window objects to single window.airportController object
* Updates airport function refs
* Abstracts airport_load to cycle through a list, adds airportLoadList.js

n8rzz committed:
Creates GameController from game.js
* Establishes gameController in App.js and adds single instance to window
* Updates references to game functions to use new window instance

n8rzz committed:
Converts tutorial to TutorialView
* Moves window functions to a single window.tutorialView instance created by App

n8rzz committed:
Converts input.js to InputController
* Adds enumsfor KEY_CODES and MOUSE_CODES
* Adds input to window via App.inputController

n8rzz committed:
Creates handlerMethods for InputController and removes .init()

n8rzz committed:
Converts ui.js to UiController
* Encapsulates ui global functions in window.uiController class method
* Adds template constants
* Updates ui function references

n8rzz commented:
Just a quick update:

Still making good progress. I've moved over to a feature branch for the last few days just to keep things isolated.  I'm almost finished tying everything together in encapsulated classes that get attached to the window.  There is still a fair amount of work to do to move everything away from global window functions.  Once I'm done on this current branch, there will be a very good base for making those changes.  It's just taking time.

I'll merge this latest work in by the end of the weekend.  yay!  another 30 or so commits for this PR.

n8rzz committed:
Adds raf and adds comments to uitl.js

n8rzz committed:
Moves canvas.js to canvas/CanvasController
* Adds constants for rgba values in CanvasController
* Encapsulates canvas methos into window.canvasController

n8rzz committed:
Adds calculateMiddle function and test. Implements new function into CanvasController

n8rzz committed:
Moves base/Position and base/Area to es6 classes
* Updates PositionModel refs to use explicit import
* Fixes some syntax errors
* Begins updating for/in with _forEach and simplifing loops

n8rzz committed:
Converts for/in loops to _forEach and simplifies loopings

n8rzz committed:
Adds many minor syntax updates

n8rzz committed:
Tears out friber from AircraftInstanceModel and AircraftFlightManagementSystem
* Updates CanvasController to use _cloneDeep isntead of $.extend()
* Removes @extends comment blocks
* Fixes typo in util.js

n8rzz committed:
Removes debugger

n8rzz committed:
Begins simplifing AircraftInstanceModel

n8rzz committed:
Abstracts AircraftStripView to a standalone class.

n8rzz committed:
Finishes AircraftStripView component. Moves local $strips to this.$strips in AircraftInstanceModel

n8rzz committed:
Begins simplification work on AircraftInstanceModel

n8rzz committed:
Abstracts more of the AircraftStrip out of AircraftInstanceModel.

n8rzz committed:
Abstracts .updateStrip() to AircraftStripView class. Siplifies logic and adds helper methods to FMS

n8rzz committed:
Updates .eslintrc with rules that were used inline. Fixes some minor lint errors.

n8rzz committed:
Removes Fiber from package.json and updates readme

n8rzz commented:
@erikquinn Fiber is officially gone!  The main refactor is done!  Load times, have improved:
<img width="281" alt="screen shot 2016-09-27 at 11 03 14 pm" src="https://cloud.githubusercontent.com/assets/3325985/18900513/72c6fa70-8507-11e6-96e6-daf44234ad9d.png">

However, work continues...

There is still a _lot_ of work to do (see updated checklist above).  I have already begun abstracting and simplifying (see: 996e567).
- Most of the main modules are still attached to the window, only now they are all encapsulated in instances that live in `App.js`.  The next step is to explicitly pass them as parameters to only the files that need them.
- As I simplify methods, I have started adding comment blocks to each method.  This is, so far, inconsistent throughout the app.  But its a start.
- `util.js` needs to be split into smaller files grouped by concern and tests need to be added.
- Various other abstractions and simplifications have been called out via `TODO` and `FIXME` comments and will be addressed soon.

When you review this, it would be easier to just pull the repo and look at the code.  I have added a readme to the `tools/` folder that walks through getting this set up locally.  I've done my best to make changes in small chunks and then test things after making those changes.  But, alas, I'm sure there are some bugs that have snuck in.  I am aware of two that are on my radar and will be addressed as part of my continuing refactor efforts.  With the amount of changes made here, `git bisect` is your best friend when zeroing in on a bug.  Seriously, know it, use it, love it.

I think this is coming along nicely, the code base is much more readable now and I'm moving towards a much better separation of concerns.  These last commits are a great example of abstraction and simplification.  

I do want to call out, if it isn't already painfully obvious, that I have touched every line of code in this project with the exception of the contents of the `parser.js` file.  That means that some file histories (git histories) are going to be lost.  Many files have been split up, renamed, or simply removed.  So be aware that the gits will say I wrote a lot of this.  I didn't.  I've just moved it around and made it look pretty.

n8rzz committed:
Adds new ignore rule to .eslintrc
* Adds new method to FMS
* Adds more todo blocks to AircraftInstanceModel

n8rzz committed:
Moves radio concerned utility functions to radioUtilities file
* Updates refs to radioUtilities functions

erikquinn commented:
Wow... At first glance, much of the new code blows my mind! :heart_eyes: But I love how the whole thing feels, it just will take some re-learning on my part. There are (of course) some gameplay/functionality issues I've come across, but nothing worrying; I'm just thrilled with all the progress! :tada:

Will have to play around with it more later, gather a high-priority hit list of stuff that broke, and then we can go bug smashing :beetle: :hammer: 

In the meantime though, unless there are important things you wanted to amend, I think it might be a decent idea to take a minute to clean up any obvious breakages, and then merge this into the `v3` branch and post a testing link, in order to give the contributors some time to help by generating a comprehensive list of issues that I may not have noticed. Then, I (or anybody) can address the bug fixes simultaneously with your continuance along your list of "upcoming tasks", thus accelerating the timeline to getting 3.0 to the big stage.

I'm getting weirdly excited by all this! haha

n8rzz commented:
Awesome, I'm glad you like what I've done. Funny thing is, even though it is laid out completely differently, _a lot_ of the old code remains mostly untouched.  Well, other than syntax changes and some simplifications here and there.  There is a healthy amount of es6 sprinkled in there, now, too.  And lodash.  Lodash makes things soooo much easier. https://lodash.com/docs/4.16.2

I've updated my list above, there are still two items that I would like to finish before we tag this as `v3.0.0` and start QAing.  Both of those tasks are in progress right now and should be done by the weekend.  After that, I'd feel comfortable with starting to squash bugs.  

How would you like to handle tracking of those bugs;  A specific tag on the existing issues board or create issues on my fork and address them there?  Its up to you, I'm not picky just so long as there is a place to collect them, track them and make comments.

ES6 has a way of making javascript fun again.  Its totally normal.  Once you start dreaming in javascript, though, then you might want to start getting concerned.  Waking up in the morning with visions of `_map(this.waypoints, (waypoint) => waypoint.name)` and such, then you may have a problem.

n8rzz commented:
One other thing, if you have any questions about what I'm doing or why (or both) please don't be afraid to ask!  I love talking javascript and I especially love talking about specific implementations.  I don't get defensive about my code and code reviews are a daily thing at my job. So I'm not shy about critiques and/or justifying the why. 

So please ask if you're curious or say something if you see something you don't like!!

n8rzz committed:
Removes random() in favor of _random() when using a min/max value
* Removes ceil() abstraction in favor of Math.ceil

n8rzz committed:
Adds assets/scripts to gh-pages branch.

n8rzz commented:
For those interested, my fork is live at: https://n8rzz.github.io/atc/

I think, just to keep things together (and @erikquinn feel free to chime in) we should log any issues with this potential `v3` codebase at: https://github.com/n8rzz/atc/issues

That way we can keep everything together, but separate.  I will continue with the last two items above and merge that into `gh-pages` once its ready.

n8rzz committed:
Uncomments GA code in index.html

n8rzz committed:
Moves vector functions to math/vector file
* Updates vector refs
* Updgrades eslint to v12.0.0
* Adds more comments to AircraftFlightManagementSystem

n8rzz committed:
Continues moving vector functions out and updating refs
* Fixes some syntax errors

n8rzz committed:
Continuing move from util to concern files. Updates eslintrc with more ignore patterns.

n8rzz committed:
Finsihed moes out of util.js
* Renames AirportInstanceModel to AirportModel
* Adds docs to App
* Deletes modules.js

n8rzz committed:
Updates anonymous functions to use arrow functions
* Removes dead code
* Adds doc blocks

n8rzz committed:
Merge branch 'develop' of https://github.com/n8rzz/atc into gh-pages

n8rzz committed:
Removes images, fonts from src/ dir.  adds styles/ to .gitignore

n8rzz committed:
Merge branch 'develop' of https://github.com/n8rzz/atc into gh-pages

n8rzz committed:
Preparation for v3.0.0 qa.

n8rzz committed:
Adds AreaModel in to AirportModel

n8rzz committed:
Merge branch 'develop' of https://github.com/n8rzz/atc into gh-pages

n8rzz committed:
Adds AreaModel

n8rzz committed:
Adds missing export default tp AreaModel

n8rzz committed:
Updates bundle

n8rzz committed:
Adds more documentation, corrects version number in package.json

n8rzz committed:
Updates bundles

n8rzz commented:
Unfortunately, I didn't get as much accomplished as I would have liked.  I was successful in moving everything in `util.js` to proper concerned function files.  There are still a few functions left in that file that may eventually move to a not-yet-created `stringHelpers.js` file.  As of right now, though, there isn't a huge need for such a file.  I did not get a chance to move the main controllers off of the `window`.  That will have to wait for another day.

I think we're at a good stopping point here to start QAing the crap out of this beast.  I have already logged a few issues that I'v found.  Mostly minor stuff, although the ILS track one is somewhat concerning.  I'll have to check the current game to see if I can reproduce it there.  

For anybody that decides to help out on bugfixes, I'd like to follow a specific branching and merging strategy:
- any bugfix should be done on a specific branch off of ~~`gh-pages`~~ `release/3.0.0` with a title `bugfix/ATC-{ISSUE_NUMBER}`.
- ~~`gh-pages` is a protected branch in my fork.~~   Once a fix is complete, please submit an PR that targets the `develop` branch.

I'm hopeful we'll be able to move through any bugfixes relatively quickly, but I've been through QA many times before and I know that that is not always the case.

At any rate, after we're done with QA I think my next order of business will be a refactor/simplification of the `AircraftInstanceModel` and the `AircraftFlightManagementSystem` along with [probably] `Legs` and `Waypoints`.  There is a lot of room for refactor in all of those files, every time I'm in them I need to restrain myself from going on a refactor binge.

Everything is up-to-date on my fork's `gh-pages` branch.  So feel free to take it for a spin (https://n8rzz.github.io/atc/) and log any issues you find (https://github.com/n8rzz/atc/issues).  If possible, please provide screenshots and steps to reproduce.  Also include your browser if you are not on Chrome.

n8rzz committed:
Updates .gitignore

n8rzz committed:
Updates .gitignore (again)

n8rzz committed:
Adds missing imports in flightMath.js

erikquinn reviewed (commented):
@n8rzz Pure curiosity: are you a sublime guy or an atom guy? Or something else altogether?

n8rzz commented:
@erikquinn yes.  And IntelliJ and Vim.

My current editor of choice is Atom.  I'm really liking it, too, it's giving vim a run for it's money.  I don't think I'll ever go back to SublimeText (2 or 3).  Too many weird bugs and not a lot of support.  Atom is great because it's easy to customize, it's built on electron and the community is very active.  Plus, being a client side developer, there is something poetic about writing code for browsers in a browser.  And it's pretty fast too.  My only complaint is with the current eslint plugins.  I have two and both are, in my opinion, sub-par.

IntelliJ is great, but it's so slow.  I was on a Scala project a while ago and IntelliJ was very useful there.  I could run the db and debug the scalas right in the IDE.  I do use it very occasionally for the merge tool and history diffing.  

Now, though, probably 85% Atom 10% Vim 5% IntelliJ.

n8rzz committed:
bugfix/ATC-5 (#7)

* bugfix/ATC-5 - refactors parseElevation to accept negative elevations
* Adds .m_ft() to convert meters to feet
* Adds m_ft ratio
* Adds tests for .parseElevation()

* Removes bundle from vcs

* Adds Airport constant to prop so that both this.airport and prop.airport point to the same reference.

n8rzz committed:
bugfix/ATC-3 - adjusts order of operation for .ui_airport_open() method. (#13)

n8rzz committed:
bugfix/ATC-4 - updates .ready() method to use a localStorage icao on load if one exists (#14)

n8rzz committed:
- bugfix/ATC-4 - updates .ready() method to use a localStorage icao on load if one exists (#14)
- feature/ATC-12 - adds http-server and `start` script to package.json that fires up a local server (#22)
- Adds comments to ArrivalBase for future refactor work
- Simplifies airline_get method and abstracts addAirline() into its own method.
- Abstracts parse logic into helper methods. Updates instances of for/in loops to use forEach instead.
- Changes .createChildren() method names to ._init()
- Adds doc blocks to new AirportModel methods.
- Abstracts if logic out of AirportModel.parse()
- Updates getFix to getFixPosition in AirportModel
- Updates references

n8rzz commented:
I forgot to add VSCode to that list as well.  I recently just started using that instead of Monodevelop  when I work with C# in Unity.

eliuuk commented:
VSCode all the way!

n8rzz committed:
feature/ATC-12 - adds http-server and `start` script to package.json that fires up a local server (#22)

n8rzz committed:
Abstracts Position to new method in AirportModel.  Adds getters for elevation and magnetic_north pulling from this.position. Adds _uniqueId to PositionModel

n8rzz committed:
Begins abstracting AirportModel.sids out to SidCollection and SidModel objects

n8rzz committed:
Adds StandardRoute collections and model objects
* Adds stubs for future tests
* Begins implementing interface with collections and models to get SID fixes

n8rzz committed:
Connects the StandardRoute interface together and returning correct array
* TODO: add tests for new classes

n8rzz committed:
bugfix/ATC-33 - changes reference to window.gameController instead of just prop.game in AircraftController. (#34)

n8rzz committed:
bugfix/ATC-20 - adds _lowerCase() to only lowercase a storage value if it exists. (#21)

n8rzz committed:
Merge branch 'develop' of https://github.com/n8rzz/atc into feature/next-version

n8rzz committed:
bugfix/ATC-23 - adjusts .change() function to use .attr() instead of .data() in UiController (#24)

* Abstracts building of each option/value set to new methods

n8rzz committed:
Finished adding doc blocks to AreaModel. Finishes tests for AreaModel.

n8rzz committed:
Merge branch 'feature/next-version' into develop

n8rzz committed:
Merge branch 'develop' of https://github.com/n8rzz/atc into feature/ATC-36

n8rzz committed:
feature/ATC-11 (#39)

* feature/ATC-11 - Adds link to tools/readme updates working in tools/readme.
* feature/ATC-11 - Adds info on adding a new airport

n8rzz committed:
Merge branch 'release/3.0.0' of https://github.com/n8rzz/atc into develop

n8rzz committed:
feature/ATC-31 - updates name and refs for arrivalFactory and departureFactory (#41)

n8rzz committed:
bugfix/ATC-1 (#8)

* Reverts _clamp to clamp and updates imports
* Adds tests for .clamp() function
* Removes bundle files
* bugfix/ATC-1 - Revert back to if statements for .clamp() after performance review of using Math turned up significant performance hit.

n8rzz committed:
Merge branch 'release/3.0.0' of https://github.com/n8rzz/atc into develop

erikquinn committed:
bugfix/ATC-2 (#15)

* Get ILS intercepting again
* Fix altitude behavior during ILS interception
* Improve localizer capture behavior
* Rename and document confusing crange/trange functions
* Code cleanup and improvements to ILS capture logic
address comments from code review

n8rzz committed:
Merge branch 'release/3.0.0' of https://github.com/n8rzz/atc into develop

n8rzz committed:
bugfix/ATC-32 - adds no cache flag to npm start cmd (#42)

n8rzz committed:
Merge branch 'release/3.0.0' of https://github.com/n8rzz/atc into develop

n8rzz committed:
Updates renamed import and fixes failing tests.

erikquinn committed:
bugfix/ATC-6 (#47)

* Fix isLanded function, rename to WOW
for Weight-On-Wheels, as it's commonly referred to
* Fix WOW logic causing issues with negative-elevation airports

n8rzz committed:
bugfix/ATC-45 (#46)

* Updates mis-matched star names

n8rzz committed:
Merge branch 'release/3.0.0' of https://github.com/n8rzz/atc into develop

n8rzz committed:
Merge branch 'develop' of https://github.com/n8rzz/atc into feature/ATC-36

n8rzz committed:
feature/ATC-36 - Renames AreaModel to AirspaceModel
* Updates references
* Moves AirspaceModel to airport forlder

n8rzz committed:
feature/ATC-36

n8rzz commented:
closing this in favor of #702
